mne/cov.py

```diff
@@ -854,6 +855,91 @@ def _unpack_epochs(epochs):
     return out
 
 
+def _merge_picks_list(picks_list):
+    """Combine grad and mag into meg picks + label."""
+    keys, picks_ = [list(ee) for ee in zip(*picks_list)]
+    picks_dict = dict(picks_list)
+    if 'mag' in keys or 'grad' in keys:
+        picks_meg = [picks_dict.get('mag', None),
+                     picks_dict.get('grad', None)]
+        picks_meg = np.concatenate(
+            [pp for pp in picks_meg if pp is not None])
+        picks_meg = np.sort(picks_meg)
+
+        picks_dict['meg'] = picks_meg
+        # now update keys to reflect merge
+        meg_index = min(keys.index(kk) if kk in keys else np.inf
+                        for kk in ('grad', 'mag'))
+        keys.insert(meg_index, 'meg')
+        for key in ('mag', 'grad'):
+            if key in keys:
+                keys.remove(key)
+
+    return [(kk, picks_dict[kk]) for kk in keys]
+
+
+def _estimate_rank_by_type(data, info, picks_list, scalings):
+    """Estimate rank from data by type."""
+    out = dict()
+    # now go over indices and compute rank
+    has_sss = False
+    if len(info['proc_history']) > 0:
+        has_sss = (info['proc_history'][0].get('max_info') is not None)
+    if has_sss:
+        picks_list_ = _merge_picks_list(picks_list)
+    else:
+        picks_list_ = picks_list
+
+    for key, this_picks in picks_list_:
+        rank = _estimate_rank_meeg_signals(
+            data,
+            pick_info(info.copy(), this_picks),
+            scalings=scalings, tol='auto', return_singular=False)
+        out[key] = rank
+
+    return out
+
+
+def _match_proj_type(proj, ch_names):
+    """See if proj should be counted."""
+    proj_ch_names = proj['data']['col_names']
+    select = any(kk in ch_names for kk in proj_ch_names)
+    return select
+
+
+def _get_expected_rank_from_info(info, picks_list):
+    """Lookup rank from info by type."""
+    from .io.proc_history import _get_rank_sss
+    out = dict()
+    # now go over indices and look up rank
+    has_sss = False
+    if len(info['proc_history']) > 0:
+        has_sss = (info['proc_history'][0].get('max_info') is not None)
+    if has_sss:
+        picks_list_ = _merge_picks_list(picks_list)
+    else:
+        picks_list_ = picks_list
+
+    for key, this_picks in picks_list_:
+        this_info = pick_info(info.copy(), this_picks)
+        rank = len(this_picks)
+        if key == 'meg' and has_sss:
+            rank = min(rank, _get_rank_sss(info))
+        n_ssp = sum(pp['active'] for pp in info['projs'] if
+                    _match_proj_type(pp, this_info['ch_names']))
```

Comment:
I don't think it depends on `active` IIRC we always apply proj at some point in the pipeline before applying the inverse

Reply:
what do you mean @larsoner? This is what I needed to get it right IIRC. Any alternative suggestion?

Reply:
I'm trying to figure out if you should just do `sum(_match_proj_type(pp, this_info['ch_names']))`. Think about the `proj='delayed'` use case or even `epochs=Epochs(proj=False)` (but will be applied later during inversion) will this code do the right thing?

Reply:
I agree with @larsoner we always end up applying all proj in the inverse to me:

sum(_match_proj_type(pp, this_info['ch_names']))

seems like the right fix